### PR TITLE
Add array of 2D textures support

### PR DESCRIPTION
**Pull request by mbouron**
Based on https://github.com/nope-project/nope.gl/pull/30.

### CODE REVIEW

File: libnopegl/src/api.c [hunk 1]
```diff
@@ -424,8 +424,10 @@ static const char *get_cap_string_id(unsigned cap_id)
     case NGL_CAP_MAX_TEXTURE_DIMENSION_2D:      return "max_texture_dimension_2d";
     case NGL_CAP_MAX_TEXTURE_DIMENSION_3D:      return "max_texture_dimension_3d";
     case NGL_CAP_MAX_TEXTURE_DIMENSION_CUBE:    return "max_texture_dimension_cube";
+    case NGL_CAP_MAX_TEXTURE_ARRAY_LAYERS:      return "max_texture_array_layers";
```
Comment (ubitux) [inline libnopegl/src/api.c:427]: alphabetic order

File: libnopegl/src/api.c [hunk 2]
```diff
@@ -469,8 +472,10 @@ static int load_caps(struct ngl_backend *backend, const struct gpu_ctx *gpu_ctx)
         CAP(NGL_CAP_MAX_TEXTURE_DIMENSION_2D,      limits->max_texture_dimension_2d),
         CAP(NGL_CAP_MAX_TEXTURE_DIMENSION_3D,      limits->max_texture_dimension_3d),
         CAP(NGL_CAP_MAX_TEXTURE_DIMENSION_CUBE,    limits->max_texture_dimension_cube),
+        CAP(NGL_CAP_MAX_TEXTURE_ARRAY_LAYERS,      limits->max_texture_array_layers),
```
Comment (ubitux) [inline libnopegl/src/api.c:475]: alphabetic order

File: libnopegl/src/backends/vk/gpu_ctx_vk.c [hunk 3]
```diff
@@ -890,6 +891,7 @@ static int vk_init(struct gpu_ctx *s)
     s->limits.max_texture_dimension_2d           = limits->maxImageDimension2D;
     s->limits.max_texture_dimension_3d           = limits->maxImageDimension3D;
     s->limits.max_texture_dimension_cube         = limits->maxImageDimensionCube;
+    s->limits.max_texture_array_layers           = limits->maxImageArrayLayers;
```
Comment (ubitux) [inline libnopegl/src/backends/vk/gpu_ctx_vk.c:894]: the full list is not in alphabetic order but for consistency you could keep it at least for the `max_texture` subset

File: libnopegl/src/gpu_limits.h [hunk 4]
```diff
@@ -38,6 +38,7 @@ struct gpu_limits {
     uint32_t max_texture_dimension_2d;
     uint32_t max_texture_dimension_3d;
     uint32_t max_texture_dimension_cube;
+    uint32_t max_texture_array_layers;
```
Comment (ubitux) [inline libnopegl/src/gpu_limits.h:41]: ditto on top of the `max_texture_*`

File: libnopegl/src/nopegl.h.in [hunk 5]
```diff
@@ -627,13 +627,15 @@ struct ngl_config {
...
 #define NGL_CAP_MAX_SAMPLES                     NGLI_FOURCC('M','S','A','A')
+#define NGL_CAP_MAX_TEXTURE_ARRAY_LAYERS        NGLI_FOURCC('M','T','A','L')
 #define NGL_CAP_MAX_TEXTURE_DIMENSION_1D        NGLI_FOURCC('M','T','D','1')
 #define NGL_CAP_MAX_TEXTURE_DIMENSION_2D        NGLI_FOURCC('M','T','D','2')
 #define NGL_CAP_MAX_TEXTURE_DIMENSION_3D        NGLI_FOURCC('M','T','D','3')
 #define NGL_CAP_MAX_TEXTURE_DIMENSION_CUBE      NGLI_FOURCC('M','T','D','C')
 #define NGL_CAP_NPOT_TEXTURE                    NGLI_FOURCC('N','P','O','T')
 #define NGL_CAP_SHADER_TEXTURE_LOD              NGLI_FOURCC('S','T','L','d')
 #define NGL_CAP_TEXTURE_3D                      NGL_NODE_TEXTURE3D
+#define NGL_CAP_TEXTURE_2D_ARRAY                NGLI_FOURCC('T','e','x','A')
```
Comment (ubitux) [inline libnopegl/src/nopegl.h.in:639]: 2D before 3D maybe (that's what you did in the rest of the diff)

File: libnopegl/src/backends/gl/texture_gl.c [hunk 6]
```diff
@@ -132,7 +132,7 @@ static void texture3d_set_sub_image(struct texture *s, const uint8_t *data, int
     if (row_upload) {
         for (int z = 0; z < params->depth; z++) {
             for (int y = 0; y < params->height; y++) {
-                ngli_glTexSubImage3D(gl, GL_TEXTURE_3D, 0, 0, y, z, params->width, 1, params->depth, s_priv->format, s_priv->format_type, data);
+                ngli_glTexSubImage3D(gl, GL_TEXTURE_3D, 0, 0, y, z, params->width, 1, 1, s_priv->format, s_priv->format_type, data);
```
Comment (ubitux) [inline libnopegl/src/backends/gl/texture_gl.c:158]: In commit message, "with row_upload is enabled", "with" → "when" or drop "is"

File: libnopegl/src/node_texture.c [hunk 7]
```diff
@@ -571,6 +630,20 @@ const struct node_class ngli_texture2d_class = {
     .file      = __FILE__,
 };
 
+const struct node_class ngli_texture2d_array_class = {
```
Comment (ubitux) [inline libnopegl/src/node_texture.c:633]: can you name this `ngli_texture2darray_class` for consistency with all the other classes?

File: libnopegl/src/node_textureview.c [hunk 8]
```diff
@@ -59,7 +59,7 @@ static int textureview_init(struct ngl_node *node)
     }
 
     if (o->texture->cls->id == NGL_NODE_TEXTURECUBE && o->layer >= 6) {
-        LOG(ERROR, "cubemap textures only have 6 layers");
+        LOG(ERROR, "requested layer (%d) exceeds cube map layer count (%d) cubemap textures only have 6 layers", o->layer, texture_params->depth);
```
Comment (ubitux) [inline libnopegl/src/node_textureview.c:62]: weird sentence break between "layer count" and "cubemap textures" 🤔

Maybe a comma or something would help?

I'm also seeing 3 variables: the requested layer, the cubemap layer count, and "6". Is `texture_params->depth` always 6 here?

File: tests/refs/texture_2d_array_from_mrt.ref [hunk 9]
```diff
@@ -0,0 +1 @@
+0:000000FF 1:404040FF 2:808080FF 3:BFBFBFFF
```
Comment (ubitux) [inline tests/refs/texture_2d_array_from_mrt.ref:1]: can you add the msaa ref as well?

**Review by ubitux — CHANGES_REQUESTED** [review]
Mostly only cosmetic comments

File: tests/refs/texture_2d_array_from_mrt.ref (hunk 9 above)
Comment (mbouron) [reply tests/refs/texture_2d_array_from_mrt.ref:1]: Done.

File: libnopegl/src/node_texture.c (hunk 7 above)
Comment (mbouron) [reply libnopegl/src/node_texture.c:633]: Done.

File: libnopegl/src/backends/gl/texture_gl.c (hunk 6 above)
Comment (mbouron) [reply libnopegl/src/backends/gl/texture_gl.c:158]: Done.

File: libnopegl/src/nopegl.h.in (hunk 5 above)
Comment (mbouron) [reply libnopegl/src/nopegl.h.in:639]: Done.

File: libnopegl/src/api.c (hunk 1 above)
Comment (mbouron) [reply libnopegl/src/api.c:427]: Done.

File: libnopegl/src/node_textureview.c (hunk 8 above)
Comment (mbouron) [reply libnopegl/src/node_textureview.c:62]: Fixed.

File: libnopegl/src/backends/vk/gpu_ctx_vk.c (hunk 3 above)
Comment (mbouron) [reply libnopegl/src/backends/vk/gpu_ctx_vk.c:894]: I choose this order to match VkPhysicalDeviceLimits, see: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceLimits.html.

File: libnopegl/src/node_textureview.c (hunk 8 above)
Comment (ubitux) [reply libnopegl/src/node_textureview.c:62]: Did you push?

Comment (mbouron) [reply libnopegl/src/node_textureview.c]: I did not, sorry. Fixed.

**Review by ubitux — APPROVED** [review]

**Pull request merged**